# QIIME2 analysis

Author: Siobhon Egan <siobhonegan@hotmail.com><siobhon.egan@murdoch.edu.au>
Last updated: Jan 20201

**Background**

This workflow is written for analysing amplicon data in QIIME2. Input data is Illumina MiSeq paired-end data prepared using Nextera XT indexes (i.e. no additional demultiplexing steps are needed in this case however should your data require demultiplexing it can easily be added in).

## 0. Install & activate QIIME2 environment (commandline)

This workflow utilsing commandline interface with QIIME2.

Requires miniconda/conda, see [here](https://docs.qiime2.org/2020.11/install/native/#installing-miniconda)

Latest version = [QIIME2-2020.11](https://docs.qiime2.org/2020.11/install/native/), see QIIME2 documentation for install based on your platform.

**Activate qiime2 environment**
```bash
conda activate qiime2-2020.11
```

To view `.qza` and `.qzv` files go to [QIIME2 view](https://view.qiime2.org/)

## 1. Input data

> Assumes paired-end data that does not require demultiplexing

Place raw data files in zipped format (i.e. `.fastq.gz` in a directory called `raw_data/`).

### File naming conventions

In [Casava 1.8 demultiplexed (paired-end)](https://docs.qiime2.org/2020.11/tutorials/importing/#casava-1-8-paired-end-demultiplexed-fastq) format, there are two `.fastq.gz` files for each sample in the study, each containing the forward or reverse reads for that sample. The file name includes the sample identifier. The forward and reverse read file names for a single sample might look like XXXX_15_L001_R1_001.fastq.gz and XXXX_15_L001_R2_001.fastq.gz, respectively. The underscore-separated fields in this file name are:

- the sample identifier,
- the barcode sequence or a barcode identifier,
- the lane number,
- the direction of the read (i.e. R1 or R2), and
- the set number.

Depending on sequencing facility you may need to add the `_001` prefix to sample files.

Note however that you do **not** need to unzip fastq data to analyse.

Navigate into the directory with raw data files:
```bash
for file in raw_data/*.fastq.gz;
do
newname=$(echo "$file" | sed 's/0_BPDNR//' | sed 's/.fastq/_001.fastq/')
mv $file $newname
done
```

### Import as QIIME2 artefact

Import `.fastq.gz` data into QIIME2 format using [Casava 1.8 demultiplexed (paired-end)](https://docs.qiime2.org/2020.11/tutorials/importing/#casava-1-8-paired-end-demultiplexed-fastq) option. Remember assumes raw data is in directory labelled `raw_data/` and file naming format as above.

```bash
qiime tools import \
--type 'SampleData[PairedEndSequencesWithQuality]' \
--input-path raw_data \
--input-format CasavaOneEightSingleLanePerSampleDirFmt \
--output-path 16S_demux_seqs.qza
```

In this case we are using Nextera Indexes which mean they are demultiplexed automatically by basespace and therefore we can skip over any reference to demultiplexing steps.

**Inspect reads for quality**
To inspect raw reads

```bash
qiime demux summarize \
  --i-data 16S_demux_seqs.qza \
  --o-visualization 16S_demux_seqs.qzv
```

View this output by importing into [QIIME2 view](https://view.qiime2.org/). Use this output to choose your parameters for QC such as trimming low quality sequences and truncating sequence length.

### Sample metadata

This holds you associated metadata related to your samples (e.g. host information, sampling data, etc). [Tutorial here](https://docs.qiime2.org/2020.11/tutorials/moving-pictures/#sample-metadata)

The metadata needs to be in `.tsv` format, the best way to do this is to access the QIIME2 googlesheet example. Save a copy and edit/add in your sample details. Then select `File > Download as > Tab-separated values`. Alternatively, the  command `wget "https://data.qiime2.org/2020.11/tutorials/moving-pictures/sample_metadata.tsv"` will download the sample metadata as tab-separated text and save it in the file `sample-metadata.tsv`. It is import you don't change the header for the first column `sample-id`.


## 2. Sequence quality control and feature table construction

> Denoise using dada2

Based on quality plot in the above output `16S_demux_seqs.qza` adjust trim length to where quality falls.

Then you can also trim primers. In this case working with 16S v1-2 data with the following primers

Example data - amplicon NGS data targetting bacteria using 16S rRNA hypervariable region 1-2 with the following primers:

- 27F-Y (20 nt): AGAGTTTGATCCTGGCTYAG #16S v1-2 primer, ref Gofton et al. Parasites & Vectors (2015) 8:345
- 338R (19 nt): TGCTGCCTCCCGTAGGAGT #16S v1-2 primer, ref Turner et al. J Eukaryot Microbiol (1999) 46(4):32

```bash
qiime dada2 denoise-paired \
  --i-demultiplexed-seqs 16S_demux_seqs.qza \
  --p-trim-left-f 20 \
  --p-trim-left-r 19 \
  --p-trunc-len-f 250 \
  --p-trunc-len-r 250 \
  --o-table 16S_denoise_table.qza \
  --o-representative-sequences 16S_denoise_rep-seqs.qza \
  --o-denoising-stats 16S_denoise-stats.qza
```

At this stage, you will have artifacts containing the feature table, corresponding feature sequences, and DADA2 denoising stats. You can generate summaries of these as follows.

```bash
qiime feature-table summarize \
  --i-table 16S_denoise_table.qza \
  --o-visualization 16S_denoise_table.qzv \
  --m-sample-metadata-file sample-metadata.tsv # Can skip this bit if needed.

qiime feature-table tabulate-seqs \
  --i-data 16S_denoise_rep-seqs.qza \
  --o-visualization 16S_denoise_rep-seqs.qzv

qiime metadata tabulate \
  --m-input-file 16S_denoise-stats.qza \
  --o-visualization 16S_denoise-stats.qzv
```

### Merging denoised artefacts

To merge denoised data sets and generate one `FeatureTable[Frequency]` and `FeatureData[Sequence]`  artifacts

```bash
qiime feature-table merge \
  --i-tables table-1.qza \
  --i-tables table-2.qza \
  --o-merged-table table.qza
qiime feature-table merge-seqs \
  --i-data rep-seqs-1.qza \
  --i-data rep-seqs-2.qza \
  --o-merged-data rep-seqs.qza
```

### Export ASV table

To produce an ASV table with number of each ASV reads per sample that you can open in excel. Use tutorial [here](https://rstudio-pubs-static.s3.amazonaws.com/489645_5fff8a6a02d84084a55e3b5b6ff960a4.html#)

Need to make biom file first

```bash
qiime tools export \
--input-path 16S_denoise_table.qza \
--output-path feature-table

biom convert \
-i feature-table/feature-table.biom \
-o feature-table/feature-table.tsv \
--to-tsv
```

### Phylogeny

Several downstream diversity metrics, available within QIIME 2, require that a phylogenetic tree be constructed using the Operational Taxonomic Units (OTUs) or Amplicon Sequence Variants (ASVs) being investigated. Documentation [here](https://docs.qiime2.org/2020.11/tutorials/phylogeny/)

```bash
qiime phylogeny align-to-tree-mafft-fasttree \
  --i-sequences rep-seqs.qza \
  --o-alignment aligned-rep-seqs.qza \
  --o-masked-alignment masked-aligned-rep-seqs.qza \
  --o-tree unrooted-tree.qza \
  --o-rooted-tree rooted-tree.qza
```

**Export**

Covert unrooted tree output to newick formatted file

```bash
qiime tools export \
  --input-path unrooted-tree.qza \
  --output-path exported-tree
```

## Taxonomy

Assign taxonomy to denoised sequences using a pre-tarined naive bayes classifier and the [q2-feature-classifier](https://docs.qiime2.org/2020.11/plugins/available/feature-classifier/) plugin. Details on how to create a classifier are available [here](link).

Note that  taxonomic classifiers perform best when they are trained based on your specific sample preparation and sequencing parameters, including the primers that were used for amplification and the length of your sequence reads.

```bash
qiime feature-classifier classify-sklearn \
--i-classifier /Taxonomy/QIIME2_classifiers_v2020.11/Silva_99_Otus/27F-388Y/classifier.qza \
--i-reads 16S_denoise_rep-seqs.qza \
--o-classification qiime2-taxa-silva/taxonomy.qza

qiime metadata tabulate \
--m-input-file qiime2-taxa-silva/taxonomy.qza \
--o-visualization qiime2-taxa-silva/taxonomy.qzv
```


In order to be able to download the sample OTU table need to do the taxonomy assignment and then make the taxa barplot. Then can download csv file with sequence number, samples and taxonomy.
[see here](https://docs.qiime2.org/2020.11/tutorials/moving-pictures/#taxonomic-analysis)

```
qiime taxa barplot \
  --i-table table.qza \
  --i-taxonomy taxonomy.qza \
  --m-metadata-file sample-metadata.tsv \
  --o-visualization taxa-bar-plots.qzv
```

Details on sample metadata available [here](https://docs.qiime2.org/2020.11/tutorials/moving-pictures/#sample-metadata)


Extra bit of code to generate a taxonomy table table to tsv from the commandline

```bash
qiime tools export \
--input-path taxonomy.qza \
--output-path exports
```

***



## Additonal optional steps


> Place to put additional steps and commands that I use from time-to-time


### 1. Trimming and primer search using `cutadapt trim-paired`

[QIIME2 link](https://docs.qiime2.org/2020.11/plugins/available/cutadapt/trim-paired/)

Search and trim for primers.

*Example*
- 16S27F-Y `AGAGTTTGATCCTGGCTYAG` ref Gofton et al. Parasites & Vectors (2015) 8:345
- 16S338R `TGCTGCCTCCCGTAGGAGT` ref Turner et al. J Eukaryot Microbiol (1999) 46(4):32

```
  qiime cutadapt trim-paired \
  --i-demultiplexed-sequences 16S_tick.qza \
  --p-adapter-f AGAGTTTGATCCTGGCTYAG \
  --p-adapter-r TGCTGCCTCCCGTAGGAGT \
  --o-trimmed-sequences primer-trimmed-16S_tick.qza
  --verbose \
  &> primer_trimming.log
```

```
  qiime demux summarize \
  --i-data primer-trimmed-16S_tick.qza \
  --o-visualization primer-trimmed-16S_tick.qzv
```

### 2. Join reads using `vsearch join-pairs`

[QIIME2 link](https://docs.qiime2.org/2020.11/tutorials/read-joining/)

If you plan to use DADA2 to join and denoise your paired end data, do not join your reads prior to denoising with DADA2; DADA2 expects reads that have not yet been joined, and will join the reads for you during the denoising process.

In QIIME 2, we use the term single-end reads to refer to forward or reverse reads in isolation; we use the term paired-end reads to refer to forward and reverse reads that have not yet been joined; and we use the term joined reads to refer to forward and reverse reads that have already been joined (or merged). It is important to understand which of these terms apply to your data, as this will determine what steps are necessary to analyze your paired-end data.

It is currently possible to join paired-end reads in QIIME 2 using the q2-vsearch plugin, or to import reads that have been joined outside of QIIME 2 (for example, with [`fastq-join`](https://github.com/brwnj/fastq-join), see [Importing pre-joined reads](https://docs.qiime2.org/2020.11/tutorials/read-joining/#importing-pre-joined-reads) for more details). This tutorial will cover both of these processes.

The input data for this is `SampleData[PairedEndSequencesWithQuality]`` artifact, which contains the demultiplexed reads (e.g. 16S_tick.qza)

[Join-pairs method in the q2-vsearch](https://docs.qiime2.org/2020.11/tutorials/read-joining/#joining-reads) plugin to join the reads:

```
qiime vsearch join-pairs \
  --i-demultiplexed-seqs demux.qza \
  --o-joined-sequences demux-joined.qza
```

You can next generate a summary of the `demux-joined.qza` artifact

```
qiime demux summarize \
  --i-data demux-joined.qza \
  --o-visualization demux-joined.qzv
```

**Output artifacts:** `demux-joined.qzv:`


Viewing a summary of joined data with read quality
You can next generate a summary of the demux-joined.qza artifact.

This summary is particularly useful for determining approximately how long your joined reads are (we’ll come back to this when we denoise with Deblur). When looking at the quality plots in this visualization, if you hover over a specific position you’ll see how many reads are at least that long (of the reads that were sampled for computing sequence quality). Make note of the highest sequence position where most (say, > 99%) of your reads are at least that long.

### 3. Sequence quality control using `quality-filter q-score`

[QIIME2 link](https://docs.qiime2.org/2020.11/tutorials/read-joining/#sequence-quality-control)
Here you can appy a apply quality control equences using [`quality-filter q-score`](https://docs.qiime2.org/2020.11/plugins/available/quality-filter/q-score/). The parameters to this method have not been extensively benchmarked on joined read data, so we recommend experimenting with different parameter settings.

```
qiime quality-filter q-score \
  --i-demux demux-joined.qza \
  --o-filtered-sequences demux-joined-filtered.qza \
  --o-filter-stats demux-joined-filter-stats.qza
```

At this stage you can choose to proceed using Deblur for additional quality control, or you can dereplicate sequences and optionally cluster them into OTUs with `q2-vsearch`. Deblur should give much higher quality results, so we recommend that procedure.

If you are instead interested in experimenting with an analysis workflow that is more like QIIME 1 processing (for example, to compare your Deblur or DADA2 result with a QIIME 1-like pipeline), you should next dereplicate and cluster your sequences. If you try this option, we strongly encourage you to call `qiime quality-filter q-score` with a higher `min-quality` threshold - possibly `--p-min-quality 20` or `--p-min-quality 30` (see Bokulich et al. 2013 for more details). You can then follow the steps in the OTU clustering tutorial. After clustering, you will likely want to filter features that are observed in only one sample using `qiime feature-table` `filter-features --p-min-samples`. In addition, removing singletons with an abundance filter is also advisable (see Bokulich et al. 2013 for more details), as well as filtering chimeric sequences.

### 4. Denoising using `deblur denoise-16S`

[QIIME2 link](https://docs.qiime2.org/2020.11/tutorials/read-joining/#deblur)

After quality filtering you can now denoise sequences with deblur. You should pass the sequence length value you selected from the quality score plots for `--p-trim-length`. This will trim all sequences to this length, and discard any sequences which are not at least this long.

```
qiime deblur denoise-16S \
  --i-demultiplexed-seqs demux-joined-filtered.qza \
  --p-trim-length 250 \
  --p-sample-stats \
  --o-representative-sequences rep-seqs.qza \
  --o-table table.qza \
  --o-stats deblur-stats.qza
```
Output artifacts:
`rep-seqs.qza`
`deblur-stats.qza`
`table.qza`

**View summary of Deblur feature table**
You can next summarize the feature table resulting from q2-deblur. This table and the corresponding representative sequences are now ready to be analyzed with the same methods and visualizers that would be used on single-end read data.

```
qiime feature-table summarize \
  --i-table table.qza \
  --o-visualization table.qzv
```
